File: src/main/java/org/keynote/godtools/android/model/Attachment.java
package org.keynote.godtools.android.model;

import android.support.annotation.Nullable;

import org.ccci.gto.android.common.jsonapi.annotation.JsonApiAttribute;
import org.ccci.gto.android.common.jsonapi.annotation.JsonApiIgnore;
import org.ccci.gto.android.common.jsonapi.annotation.JsonApiType;
import org.keynote.godtools.android.util.ModelUtils;

import static org.keynote.godtools.android.model.Attachment.JSON_API_TYPE;

@JsonApiType(JSON_API_TYPE)
public class Attachment extends Base {
    static final String JSON_API_TYPE = "attachment";
    private static final String JSON_RESOURCE = "resource";
    private static final String JSON_FILE_NAME = "file-file-name";
    private static final String JSON_SHA256 = "sha256";

    @Nullable
    @JsonApiIgnore
    private Long mToolId;
    @Nullable
    @JsonApiAttribute(name = JSON_RESOURCE)
    private Tool mTool;

    @Nullable
    @JsonApiAttribute(name = JSON_FILE_NAME)
    private String mFileName;
    @Nullable
    @JsonApiAttribute(name = JSON_SHA256)
    private String mSha256;

    @JsonApiIgnore
    private boolean mDownloaded = false;

    public long getToolId() {
        return mToolId != null && mToolId != Tool.INVALID_ID ? mToolId :
                mTool != null ? mTool.getId() : Tool.INVALID_ID;
    }

    public void setToolId(@Nullable final Long id) {
        mToolId = id;
    }

    @Nullable
    public String getFileName() {
        return mFileName;
    }

    public void setFileName(@Nullable final String name) {
        mFileName = name;
    }

    @Nullable
    public String getSha256() {
        return mSha256;
    }

    public void setSha256(@Nullable final String sha256) {
        mSha256 = sha256;
    }

    @Nullable
    public String getLocalFileName() {
        return ModelUtils.getAttachmentLocalFileName(mSha256, mFileName);
    }

    public boolean isDownloaded() {
        return mDownloaded;
    }

    public void setDownloaded(final boolean state) {
        mDownloaded = state;
    }
}

File: library/model/src/main/java/org/cru/godtools/model/Attachment.kt
package org.cru.godtools.model

import org.ccci.gto.android.common.jsonapi.annotation.JsonApiAttribute
import org.ccci.gto.android.common.jsonapi.annotation.JsonApiIgnore
import org.ccci.gto.android.common.jsonapi.annotation.JsonApiType
import org.cru.godtools.base.FileManager

private const val JSON_API_TYPE = "attachment"
private const val JSON_RESOURCE = "resource"
private const val JSON_FILE_NAME = "file-file-name"
private const val JSON_SHA256 = "sha256"

@JsonApiType(JSON_API_TYPE)
class Attachment : Base() {
    @JsonApiAttribute(JSON_RESOURCE)
    private val tool: Tool? = null
    @JsonApiIgnore
    var toolId: Long? = null
        get() = field?.takeUnless { it == INVALID_ID } ?: tool?.id ?: INVALID_ID

    @JsonApiAttribute(JSON_FILE_NAME)
    var filename: String? = null
    @JsonApiAttribute(JSON_SHA256)
    var sha256: String? = null

    @JsonApiIgnore
    var isDownloaded = false

    val localFilename: String?
        get() = sha256?.let { sha256 ->
            val extension = filename?.substringAfterLast('.', "bin") ?: "bin"
            "$sha256.$extension"
        }

    suspend fun getFile(manager: FileManager) = localFilename?.let { manager.getFile(it) }
    fun getFileBlocking(manager: FileManager) = localFilename?.let { manager.getFileBlocking(it) }
}
